File: Wave/compiler.c
#include "compiler.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

void process_line(char* line, FILE* output_file);

// Helper function to determine if a string represents an integer
int is_integer(const char* str) {
    while (*str) {
        if (!isdigit(*str) && *str != '-') {
            return 0;
        }
        str++;
    }
    return 1;
}

void handle_import(const char* module_name, FILE* output_file) {
    char wave_filename[300];
    snprintf(wave_filename, sizeof(wave_filename), "examples/%s.wave", module_name);

    // printf("Importing module: %s\n", wave_filename);  // Debug message

    FILE* import_file = fopen(wave_filename, "r");
    if (!import_file) {
        perror("Failed to open imported module");
        exit(EXIT_FAILURE);
    }

    char import_line[256];
    while (fgets(import_line, sizeof(import_line), import_file)) {
        // printf("Processing line from import: %s", import_line);  // Debug message
        process_line(import_line, output_file);
    }

    fclose(import_file);
}

void process_line(char* line, FILE* output_file) {
    // printf("Processing line: %s", line);  // Debug message

    // Trim newline characters
    line[strcspn(line, "\r\n")] = '\0';

    if (strstr(line, "fun")) {
        fprintf(output_file, "void ");
        char* start = strstr(line, "fun") + 4;
        char* end = strchr(start, '{');
        if (end) *end = '\0';
        fprintf(output_file, "%s {\n", start);
    }
    else if (strstr(line, "var")) {
        // Handle variable declarations
        char* start = strstr(line, "var") + 4;
        char* eq_pos = strchr(start, '=');
        if (eq_pos) {
            *eq_pos = '\0'; // Split the variable name and the value
            char* name = start;
            char* value = eq_pos + 1;

            // Determine variable type and format accordingly
            if (is_integer(value)) {
                fprintf(output_file, "int %s = %s;\n", name, value);
            }
            else if (value[0] == '"') {
                // Handle strings (assuming double quotes for string literals)
                fprintf(output_file, "char %s[] = %s;\n", name, value);
            }
            else {
                fprintf(output_file, "int %s = %s;\n", name, value); // Default to int if unknown
            }
        }
    }
    else if (strstr(line, "import")) {
        char module_name[256];
        sscanf_s(line, "import %s", module_name);
        handle_import(module_name, output_file);
    }
    else if (strstr(line, "if") || strstr(line, "else") || strstr(line, "while")) {
        fprintf(output_file, "%s\n", line);
    }
    else {
        fprintf(output_file, "%s\n", line);
    }
}

void compile_wave_to_c(const char* input_filename, const char* output_filename) {
    FILE* input_file = fopen(input_filename, "r");
    FILE* output_file = fopen(output_filename, "w");

    if (!input_file || !output_file) {
        perror("Error opening files");
        exit(EXIT_FAILURE);
    }

    fprintf(output_file, "#include <stdio.h>\n\n");

    char line[256];
    while (fgets(line, sizeof(line), input_file)) {
        process_line(line, output_file);
    }

    fclose(input_file);
    fclose(output_file);
}

void compile_and_run(const char* c_filename) {
    char command[256];
    snprintf(command, sizeof(command), "gcc %s -o wave_output && wave_output", c_filename);
    system(command);
}

File: src/compiler.rs
// compiler.rs
use std::fs::File;
use std::io::{self, BufRead, Write};
use std::path::Path;
use std::process::Command;

fn is_integer(s: &str) -> bool {
    s.chars().all(|c| c.is_digit(10) || c == '-')
}

fn handle_import(module_name: &str, output_file: &mut File) -> io::Result<()> {
    let wave_filename = format!("examples/{}.wave", module_name);
    let import_file = File::open(&wave_filename)?;

    let reader = io::BufReader::new(import_file);
    for line in reader.lines() {
        let line = line?;
        process_line(&line, output_file)?;
    }

    Ok(())
}

fn process_line(line: &str, output_file: &mut File) -> io::Result<()> {
    let line = line.trim_end_matches(&['\r', '\n'] as &[_]);

    if line.contains("fun") {
        writeln!(output_file, "void {} {{", &line[4..line.find('{').unwrap_or(line.len())])?;
    } else if line.contains("var") {
        let parts: Vec<&str> = line[4..].split('=').collect();
        if parts.len() == 2 {
            let name = parts[0].trim();
            let value = parts[1].trim();

            // 여기에서 var를 삭제하고 int로만 변경합니다.
            if is_integer(value) {
                writeln!(output_file, "int {} = {};", name, value)?; // 'var'를 'int'로 수정
            } else if value.starts_with('"') {
                writeln!(output_file, "char {}[] = {};", name, value)?;
            } else {
                writeln!(output_file, "int {} = {};", name, value)?;
            }
        }
    } else if line.contains("import") {
        let module_name = line.split_whitespace().nth(1).unwrap_or("");
        handle_import(module_name, output_file)?;
    } else if line.contains("if") || line.contains("else") || line.contains("while") {
        writeln!(output_file, "{}", line)?;
    } else {
        writeln!(output_file, "{}", line)?;
    }

    Ok(())
}

pub fn compile_wave_to_c(input_filename: &str, output_filename: &str) -> io::Result<()> {
    let input_file = File::open(input_filename)?;
    let mut output_file = File::create(output_filename)?;

    writeln!(output_file, "#include <stdio.h>\n")?;

    let reader = io::BufReader::new(input_file);
    for line in reader.lines() {
        let line = line?;
        process_line(&line, &mut output_file)?;
    }

    Ok(())
}

pub fn compile_and_run(c_filename: &str) -> io::Result<()> {
    Command::new("gcc")
        .arg(c_filename)
        .arg("-o")
        .arg("wave_output")
        .status()?;

    Command::new("./wave_output").status()?;

    Ok(())
}
